python/tvm/autotvm/task/space.py

```diff
@@ -175,51 +189,60 @@ def __init__(self, axes, policy, **kwargs):
         self.policy = policy
         self.entities = []
 
-        if policy == 'all':
-            num_outputs = kwargs["num_outputs"]
-            max_factor = kwargs.get("max_factor", 1 << 31)
-            fil = kwargs.get("filter", lambda x: True)
-
-            length = axis.length
-            factors = get_factors(length)
-            factors = [x for x in factors if x <= max_factor]
-            # copy factors for every level
-            self.product = length
-            self.num_outputs = num_outputs
-            self.factors = [factors] * (num_outputs-1)
-            self._generate_space(0, [None] * (num_outputs - 1))
-            self.entities = list(filter(fil, self.entities))
-            self.num_output = num_outputs
-        elif policy == 'candidate':
-            self.product = axis.length
-            self.num_outputs = kwargs["num_outputs"]
+        max_factor = kwargs.get("max_factor", 1 << 31)
+        fil = kwargs.get("filter", lambda x: True)
+        self.product = axis.length
+        self.num_output = kwargs.get("num_outputs", 0)
+        assert self.num_output > 0
+
+        if policy == 'candidate':
             for size in kwargs["candidate"]:
-                assert len(size) == self.num_outputs
-                # assert np.prod(size) == self.product
+                assert len(size) == self.num_output
                 self.entities.append(SplitEntity(size))
-            self.num_output = self.num_outputs
         else:
-            raise RuntimeError("Invalid policy: " + policy)
+            if policy == 'verbose':
+                # Include factors and power-of-twos. May generate tails.
+                divisibles = get_factors(self.product)
+                pow2s = get_pow2s(self.product)
+                factors = [x for x in list(set(divisibles) | set(pow2s)) if x <= max_factor]
+            elif policy == 'factors':
+                # Include divisible factors. Guarantee no tails.
+                factors = [x for x in get_factors(self.product) if x <= max_factor]
+            elif policy == 'power2':
+                # Include less, equal, and round-up power-of-two numbers. May generate tails.
+                factors = [x for x in get_pow2s(self.product) if x <= max_factor]
+            else:
+                raise RuntimeError("Invalid policy: %s" % policy)
 
-    def _generate_space(self, now, tmp_stack):
+            # Enforce the product of all split factors equals to the axis length
+            no_tail = kwargs.get("no_tail", policy == 'factors')
```

Comment:
document `no_tail` arg in `define_split`

Reply:
thanks for the remind.